File: EARestrictedScrollView/EARestrictedScrollView.m
//
//  EARestrictedScrollView.m
//
//  Copyright (c) 2015 Evgeny Aleksandrov. License: MIT.

#import "EARestrictedScrollView.h"

@interface EARestrictedScrollView ()

@property(nonatomic, assign) CGPoint containerOffset;

@property(nonatomic, strong) UIView *containerView;

@end

@implementation EARestrictedScrollView

#pragma mark - Subviews override

- (void)addSubview:(UIView *)view {
    if([self.subviews count] < 3 && [self checkIfScrollIndicator:view]) {
        [super addSubview:view];
    } else {
        [self.containerView addSubview:view];
    }
}

- (void)insertSubview:(UIView *)view aboveSubview:(UIView *)siblingSubview {
    [self.containerView insertSubview:view aboveSubview:siblingSubview];
}

- (void)insertSubview:(UIView *)view atIndex:(NSInteger)index {
    [self.containerView insertSubview:view atIndex:index];
}

- (void)insertSubview:(UIView *)view belowSubview:(UIView *)siblingSubview {
    [self.containerView insertSubview:view belowSubview:siblingSubview];
}

- (void)bringSubviewToFront:(UIView *)view {
    if(view.superview == self) {
        [super bringSubviewToFront:view];
    } else {
        [self.containerView bringSubviewToFront:view];
    }
}

- (void)sendSubviewToBack:(UIView *)view {
    [self.containerView sendSubviewToBack:view];
}

- (UIView *)viewWithTag:(NSInteger)tag {
    return [self.containerView viewWithTag:tag];
}

- (NSArray *)subviews {
    return self.containerView.subviews;
}

#pragma mark - Private checks

- (BOOL)checkIfScrollIndicator:(UIView *)view {
    return ((self.showsHorizontalScrollIndicator && view.frame.size.height == 2.5f) || (self.showsVerticalScrollIndicator && view.frame.size.width == 2.5f)) && [view isKindOfClass:[UIImageView class]];
}

#pragma mark - Lazy properties

- (UIView *)containerView {
    if(!_containerView || ![_containerView superview]) {
        _containerView = [[UIView alloc] init];
        [super addSubview:_containerView];
    }
    
    return _containerView;
}

#pragma mark - Custom offset getters and setters

- (CGPoint)contentOffset {
    CGPoint originalOffset = [super contentOffset];
    CGPoint newOffset = CGPointMake(originalOffset.x - self.containerOffset.x, originalOffset.y - self.containerOffset.y);
    
    return newOffset;
}

- (void)setContentSize:(CGSize)contentSize {
    [super setContentSize:contentSize];
    
    [self.containerView setFrame:CGRectMake(self.containerView.frame.origin.x, self.containerView.frame.origin.y, contentSize.width, contentSize.height)];
}

- (void)setRestrictionArea:(CGRect)restrictionArea {
    _restrictionArea = restrictionArea;
    
    [self.containerView setFrame:CGRectMake(-restrictionArea.origin.x, -restrictionArea.origin.y, restrictionArea.size.width, restrictionArea.size.height)];
    [super setContentSize:CGSizeMake(restrictionArea.size.width, restrictionArea.size.height)];
}

@end

File: Source/EARestrictedScrollView.swift
//
//  EARestrictedScrollView.swift
//
//  Copyright (c) 2015-2016 Evgeny Aleksandrov. License: MIT.

import UIKit

open class EARestrictedScrollView: UIScrollView {
    
    /// Container to hold all subviews of scrollview.
    lazy fileprivate var containerView: UIView = self.createContainerView()
    
    /// Helper func, since direct use of `super` call in `lazy` causes compile error.
    func createContainerView() -> UIView {
        let view = UIView()
        super.addSubview(view)
        return view
    }
    
    /// Affects `restrictionArea.size` and `containerView.frame.size` when set.
    override open var contentSize: CGSize {
        get {
            return super.contentSize
        }
        set(newContentSize) {
            containerView.frame = CGRect(origin: containerView.frame.origin, size: newContentSize)
            restrictionArea = CGRect(origin: restrictionArea.origin, size: newContentSize)
        }
    }
    
    /// Recalculated `contentOffset` in coordinate space of `containerView`.
    open var alignedOffset: CGPoint {
        get {
            let originalOffset = super.contentOffset
            let newOffset = CGPoint(x: originalOffset.x + restrictionArea.origin.x, y: originalOffset.y + restrictionArea.origin.y)
            
            return newOffset
        }
        set(newContentOffset) {
            let newOffset = CGPoint(x: newContentOffset.x - restrictionArea.origin.x, y: newContentOffset.y - restrictionArea.origin.y)
            
            super.contentOffset = newOffset
        }
    }
    
    /// Defines restriction area in coordinate space of `containerView`. Use CGRectZero to reset restriction.
    open var restrictionArea: CGRect = CGRect.zero {
        didSet {
            if restrictionArea == CGRect.zero {
                super.contentOffset = CGPoint(x: super.contentOffset.x - containerView.frame.origin.x, y: super.contentOffset.y - containerView.frame.origin.y)
                containerView.frame = CGRect(origin: CGPoint.zero, size: containerView.frame.size)
                super.contentSize = containerView.frame.size
            } else {
                containerView.frame = CGRect(origin: CGPoint(x: -restrictionArea.origin.x, y: -restrictionArea.origin.y), size: containerView.frame.size)
                super.contentOffset = CGPoint(x: super.contentOffset.x - restrictionArea.origin.x, y: super.contentOffset.y - restrictionArea.origin.y)
                super.contentSize = restrictionArea.size
            }
        }
    }
    
    /// Leads to `containerView.subviews` - all subviews except scroll indicators are stored there.
    open var containedSubviews: [UIView] {
        return containerView.subviews
    }
    
    // MARK: - Subviews functions override
    
    override open func addSubview(_ view: UIView) {
        if subviews.count < 3 && isItScrollIndicator(view) {
            super.addSubview(view)
        } else {
            containerView.addSubview(view)
        }
    }
    
    override open func insertSubview(_ view: UIView, aboveSubview siblingSubview: UIView) {
        containerView.insertSubview(view, aboveSubview: siblingSubview)
    }
    
    override open func insertSubview(_ view: UIView, at index: Int) {
        containerView.insertSubview(view, at: index)
    }
    
    override open func insertSubview(_ view: UIView, belowSubview siblingSubview: UIView) {
        containerView.insertSubview(view, belowSubview: siblingSubview)
    }
    
    override open func bringSubview(toFront view: UIView) {
        if view.superview == self {
            super.bringSubview(toFront: view)
        } else {
            containerView.bringSubview(toFront: view)
        }
    }
    
    override open func sendSubview(toBack view: UIView) {
        if view.superview == self {
            super.sendSubview(toBack: view)
        } else {
            containerView.sendSubview(toBack: view)
        }
    }
    
    // MARK: - Private
    
    /// Checks if a view is a scroll indicator of `UIScrollView`.
    fileprivate func isItScrollIndicator(_ view: UIView) -> Bool {
        return ((showsHorizontalScrollIndicator && view.frame.height == 2.5) || (showsVerticalScrollIndicator && view.frame.width == 2.5)) && view is UIImageView
    }
}
